### PR TITLE
feat: mcp tool allowlist

### PR DESCRIPTION
**Pull request by alienzach**
### TL;DR

Added the ability to filter which tools are exposed from an MCP server to agents.

### What changed?

- Added a new `allowed_tools` configuration option to `MCPServerSettings` that accepts a set of tool names
- Modified the server loading logic in `mcp_aggregator.py` to filter tools based on the `allowed_tools` configuration
- When `allowed_tools` is specified, only tools with names that exactly match entries in the set will be exposed to agents

### How to test?

1. Configure a server with the `allowed_tools` option in your configuration:
   ```yaml
   servers:
     my_server:
       allowed_tools: 
         - "tool_one"
         - "tool_two"
   ```
2. Start the MCP agent with this configuration
3. Verify that only the specified tools are available to agents
4. Try accessing a tool that's not in the allowed list and confirm it's not available

### Why make this change?

This change provides more granular control over which tools are exposed to agents from each server. This is useful for:
- Security purposes - limiting access to potentially dangerous tools
- Simplifying the agent experience by only exposing relevant tools
- Creating different tool configurations for different agent scenarios without needing separate server instances

## Summary by CodeRabbit

* **New Features**
  * Per-server allow-list to restrict which tools agents can see (exact name matching).
  * New configuration option for ID reuse policy to control temporal task ID behavior.

* **Behavior**
  * Discovery filters tools so only permitted tools are indexed and shown; an explicitly empty allow-list exposes no tools.
  * Metrics/reporting include a per-server filtered-tools count.

* **Tests**
  * Added comprehensive tests covering per-server tool filtering and edge cases.

Resolving: https://github.com/lastmile-ai/mcp-agent/issues/183, https://github.com/lastmile-ai/mcp-agent/issues/206

Related: https://github.com/lastmile-ai/mcp-agent/pull/236

### CODE REVIEW

**Comment by alienzach** [comment]
* **#413** <a href="https://app.graphite.dev/github/pr/lastmile-ai/mcp-agent/413?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#411** <a href="https://app.graphite.dev/github/pr/lastmile-ai/mcp-agent/411?utm_source=stack-comment-icon" target="_blank">Graphite</a> 👈 <a href="https://app.graphite.dev/github/pr/lastmile-ai/mcp-agent/411?utm_source=stack-comment-view-in-graphite" target="_blank">(View in Graphite)</a>
* `main`


<h2></h2>

This stack of pull requests is managed by <a href="https://graphite.dev?utm-source=stack-comment"><b>Graphite</b></a>. Learn more about <a href="https://stacking.dev/?utm_source=stack-comment">stacking</a>.

**Comment by alienzach** [comment]
Resolving: https://github.com/lastmile-ai/mcp-agent/issues/183, https://github.com/lastmile-ai/mcp-agent/issues/206

Related: https://github.com/lastmile-ai/mcp-agent/pull/236

File: src/mcp_agent/mcp/mcp_aggregator.py
Comment (saqadri) [inline src/mcp_agent/mcp/mcp_aggregator.py:1]: You'll need to run `make schema` to regenerate the config schema and include that in the PR

**Review by saqadri — CHANGES_REQUESTED** [review]
Generally looks good -- can you please also add tests for this new logic as part of your PR? Second, it would be great for someone to specify overrides for this in `RequestParams`. That logic will have to be managed in `augmented_llm.py`. Can add that as part of this PR or subsequent PR, but tests should be in this same PR.

File: tests/mcp/test_mcp_aggregator.py [hunk 1]
```diff
@@ -860,3 +861,320 @@ async def get_prompt(self, name, arguments=None):
     assert (
         hasattr(result, "description") and "Error getting prompt" in result.description
     )
+
+
+# =============================================================================
+# Tool Filtering Tests
+# =============================================================================
+
+# MockTool removed - using proper MCP Tool class instead
```
Comment (saqadri) [inline tests/mcp/test_mcp_aggregator.py]: Remove
```suggestion
```
